### PR TITLE
[CMAKE] Add circt library declaration helpers

### PR DESCRIPTION
**Pull request by youngar**
This change adds many cmake helpers to declare circuit libraries, for
example:

`cmake
add_circt_dialect(RTL rtl)
add_circt_library(RTL ...)
`

These new functions are implemented by thinly wrapping the MLIR version
of these calls, doing some CIRCT specific logic in the process.

The intention is that eventually these functions will be replaced with
an implementation that does not wrap `add_mlir_library()`. If CIRCT
migrates in to the LLVM main repo, these will have to be properly
implemented without hijacking the MLIR cmake logic.

It seems that the functionality implemented by these functions is
required by any new MLIR based project.  In the future we should look
into refactoring them into common project helpers.  If a re-usable
framework of core utilities is created in LLVM, we will switch to using
those.

### CODE REVIEW

**Comment by youngar** [comment]
See https://github.com/llvm/circt/issues/297 for more details

File: cmake/modules/CMakeLists.txt
Comment (stephenneuendorffer) [inline cmake/modules/CMakeLists.txt:21]: I don't think you have anything that uses these?

File: cmake/modules/AddCIRCT.cmake
Comment (stephenneuendorffer) [inline cmake/modules/AddCIRCT.cmake:9]: spelling?

File: cmake/modules/AddCIRCT.cmake
Comment (stephenneuendorffer) [inline cmake/modules/AddCIRCT.cmake:19]: To me, the point of this is the have add_circt_library defer to add_llvm_library, not add_mlir_library().  is that your eventual intention?

File: cmake/modules/AddCIRCT.cmake
Comment (stephenneuendorffer) [inline cmake/modules/AddCIRCT.cmake:6]: It gives me the heebie-jeebies to see this stuff, which is going to get replicated in many many downstream MLIR projects.  I'm not opposed to adding this temporarily, but I'd really rather see it get factored out so that any MLIR-dependent project can instantiate this boilerplate.

File: CMakeLists.txt
Comment (stephenneuendorffer) [inline CMakeLists.txt:76]: Are these changes relating to include paths separable from the AddCIRCT changes?

**Review by stephenneuendorffer — APPROVED** [review]
This is mostly good progress, I don't mind seeing this go in as part of incremental improvements, as long as more is forthcoming :)

File: cmake/modules/CMakeLists.txt
Comment (youngar) [reply cmake/modules/CMakeLists.txt:21]: These need to be defined in the lexical environment so that they are substituted into `CIRCTConfig.cmake.in` during the call to `configure_file`. I can't find a better way to substitute properties, but I'm no CMake wizard. This is taken from MLIR.

File: CMakeLists.txt
Comment (youngar) [reply CMakeLists.txt:76]: The most important bit is changing what `CIRCT_BINARY_DIR` is set to.  This is needed to have the correct include directories in `CIRCTConfig.cmake`.  Without that change, the include directory is set to `build/bin/include` instead of `build/include`.

The line:
```cmake
set(CIRCT_CONFIG_INCLUDE_DIRS
  "${CIRCT_SOURCE_DIR}/include"
  "${CIRCT_BINARY_DIR}/include"
  )
```

leads to:
```cmake
set(CIRCT_INCLUDE_DIRS "/Users/andrewy/wsp/circt/include;/Users/andrewy/wsp/circt/build/bin/include")
```

This can probably be separated out into another change, as no one is relying on any of this anyways. Should I go ahead an do this?

File: cmake/modules/AddCIRCT.cmake
Comment (youngar) [reply cmake/modules/AddCIRCT.cmake:19]: I guess the main point of this change was calling the right functions throughout our project, and getting our targets properly exported from `CIRCTConnfig.cmake`.  I was hoping to keep this as a largely non-functional change to how the libraries are built.

I thought replacing `add_mlir_library()` would have been a larger change, best saved for later.  It does quite a bit of stuff, a lot of which I do not quite understand yet https://github.com/llvm/llvm-project/blob/main/mlir/cmake/modules/AddMLIR.cmake#L51

Comment (stephenneuendorffer) [reply cmake/modules/AddCIRCT.cmake:19]: Makes sense.  It would be good to capture this in the commit comments.

File: CMakeLists.txt
Comment (stephenneuendorffer) [reply CMakeLists.txt:76]: I think it would make things easier to understand.  Also a good thing to capture in the commit coments.

File: cmake/modules/CMakeLists.txt
Comment (stephenneuendorffer) [reply cmake/modules/CMakeLists.txt:21]: Right...  My only concern is that this might encourage people to use the environment variable rather than the property and they might eventually get out of sync.  In MLIR, the property is used as the golden source of truth and only extracted in the places when needed, IIRC.

**Comment by stephenneuendorffer** [comment]
Thanks for pushing this forward!

**Review by stephenneuendorffer — APPROVED** [review]

File: cmake/modules/CMakeLists.txt
Comment (youngar) [reply cmake/modules/CMakeLists.txt:21]: I hope I'm getting this right:

The environment variable I'm creating is only accessible in `cmake/modules/CMakeLists.txt` and any sub-directories (there are none).  If this was in the top level CMakeLists.txt, then I could see there being a problem.  The `GLOBAL` specifier here is to specify that the property should be loaded from global properties (as opposed to a specific target's property).

Comment (youngar) [reply cmake/modules/CMakeLists.txt:21]: I can separate the stuff related to exporting in to a separate PR.  I did all the work at the same time to make sure that what I was doing worked.  How does that sound?

Comment (stephenneuendorffer) [reply cmake/modules/CMakeLists.txt:21]: Yes, you're right about the scoping.  I thought this was in the toplevel CMakeLists.txt.  This is fine as it is.

**Comment by youngar** [comment]
I split out some of the changes related to CIRCTConfig.cmake in to another PR: https://github.com/llvm/circt/pull/306

File: CMakeLists.txt
Comment (youngar) [reply CMakeLists.txt]: These have been split in to another pr

**Comment by youngar** [comment]
Hey @stephenneuendorffer, I want to check if you are ok with this change.  Flushing out the implementation of these functions is not the highest priority on my list, so it could take a while to get these fixed. I could hold off merging something like this until there are more concrete plans, but it would mean our CIRCTConfig.cmake is still totally broken.

**Comment by stephenneuendorffer** [comment]
I think it makes sense to commit this and it will be there when we are ready to build on it.  Thanks Andrew!

**Comment by youngar** [comment]
@stephenneuendorffer Thanks for your help with this!

**Pull request merged**
